### PR TITLE
Increase effect of mime's vow of silence

### LINKED ISSUE #25796: Mute characters should not snore when they sleep
Title.

### PR DESCRIPTION
**Pull request by Tayrtahn**
## About the PR

Mimes with an active vow of silence now have more silencing effects applied to them.

Specifically,
- Mimes no longer snore while sleeping.
- Mimes now eat and drink silently.
- ~Mimes no longer make footstep sounds.~
- ~Mimes can no longer make sounds with a variety of sound-making objects, such as whistles, horns, plushies, desk bells, whoopie cushions, and more. Any non-sound effects of these items still occur (whistles still make ! show up).~

Breaking their vow will cancel these effects until they retake it.

These effects are tied to the mime's power specifically, so characters who are mute for other reasons do not have them.

Fixes #25796.

## Why / Balance

The mime's vow is powerful magic! Mimes shouldn't be able to cheat by making sounds other ways.

Yes, a lot of this is rather silly, but I feel like it fits the style of Space Station. This adds some more character to the mime makes it a more unique role.

These effects can be turned on or off by a boolean property on SilencedComponent, so they can be reused on other entities if desired.

## Technical details

MimePowersSystem now adds a Silenced component to characters alongside the Muted component.

SilencedSystem subscribes to some newly added events on SleepingSystem, FoodSystem, and DrinkSystem.

SharedEmitSoundSystem.TryEmitSound now returns a boolean indicating if the sound was able to be played or not. The server's EmitSoundSystem now uses this to prevent displaying the optional popup message when playing a sound in the event that the sound doesn't play. This isn't used by the current implementation, but it's probably worth keeping. 

SharedSleepingSystem was in the Server namespace instead of Shared, so that was corrected as well.

## Media

My screen recording software doesn't want to record audio, so there's not a lot I can show here.

- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

SharedSleepingSystem is now actually in the Shared namespace.

**Changelog**

:cl:
- tweak: The mime's vow of silence now silences them while eating, drinking, and sleeping.

### CODE REVIEW

**Comment by UbaserB** [comment]
If the mime has the trait ‘mute’ AND is under the vow of silence, i am assuming these changes are still applied?

**Comment by Tayrtahn** [comment]
> If the mime has the trait ‘mute’ AND is under the vow of silence, i am assuming these changes are still applied?

Yeah, there's nothing that _prevents_ mute characters from being silenced by this, it's just an additional effect applied by the mime's power.

**Comment by lzk228** [comment]
> * Mimes no longer make footstep sounds.

isnt that too OP

**Comment by Tayrtahn** [comment]
> > * Mimes no longer make footstep sounds.
> 
> isnt that too OP

Possibly. It's definitely a buff to syndie mimes. Whether it's enough to be a balance problem is worth discussing.

**Comment by Everturning** [comment]
I dislike the footsteps and the plushies change. Mimes shouldn't be like ghostly spectres, because they're PEOPLE. The laws of sound still apply to people and plushies, even if said people can make invisible walls. 

Whistle change is funny though, keep it in. Other than that, i see no reason why anything other than that shouldn't make sound if it isn't connected to the mime's vocals. IRL, the only thing that's silent about mimes is their mouth, but i can see why this would be exaggerated. Even so, they shouldn't be completely incapable of sound whatsoever.

**Comment by PolterTzi** [comment]
> I dislike the footsteps and the plushies change. Mimes shouldn't be like ghostly spectres, because they're PEOPLE. The laws of sound still apply to people and plushies, even if said people can make invisible walls.

I mean, we agree mimes have magic powers that can violate the laws of physics.
I don't think breaking the laws of physics is a good argument here.

**Comment by SlamBamActionman** [comment]
I think from a mechanical standpoint I'd rather see silent footsteps implemented elsewhere rather than be a mime trait. It's a very interesting and potentially powerful ability, of which mimes already have one (invisible wall). Would be more fun as a Syndie item or similar (slowwalking while wearing socks/no shoes, perhaps?).

It also feels a bit uninspired. Mimes are silent, yes, but they interact with the physical world. Just going "haha mime no sound" for the job's design isn't particularly interesting.

**Comment by deltanedas** [comment]
tg has an advanced mimery guide or something  the footsteps part could be moved to that whenever implemented so sus mimes arent as op

**Comment by Tayrtahn** [comment]
> tg has an advanced mimery guide or something  the footsteps part could be moved to that whenever implemented so sus mimes arent as op

That sounds reasonable to me. I'll make the footstep silencing off by default and leave the functionality for the future.

**Comment by K-Dynamic** [comment]
> I think from a mechanical standpoint I'd rather see silent footsteps implemented elsewhere rather than be a mime trait. 

Just make walking for everyone suppress footstep sounds by like 80%

**Comment by UbaserB** [comment]
> > tg has an advanced mimery guide or something  the footsteps part could be moved to that whenever implemented so sus mimes arent as op
> 
> That sounds reasonable to me. I'll make the footstep silencing off by default and leave the functionality for the future.

instead of being muted, can you simply make it quieter? I thought it was kind of cool.

**Comment by Krunklehorn** [comment]
image

It may sound oxymoronic, but there are Mime players for whom sound is integral to their character's appeal or integral to the way they communicate with others.

Mimes can't speak and that's enough of a challenge to be interesting. Removing sound from certain interactions just stifles creativity. Picking up a bike horn only to find out it makes no noise because 'Mime Powers' is just frustrating.

**Comment by deltanedas** [comment]
also make mime using clown/cluwne horn instantly break vow for being a traitor to all mimes :trollface:

**Comment by Tayrtahn** [comment]
As much as the mental image of a mind honking a bike horn IRL seems wrong to me, I think I'll drop that aspect of this PR due to the negative reactions from people who actually enjoy playing mime in the game.

Maybe it could be a cvar for "hardcore mimes" in case downstreams or other servers want to use it.

**Comment by Tayrtahn** [comment]
Alright, footsteps and sound emitters make noise by default now. Either can be changed separately by server-only CVars (mime.allow_footsteps and mime.allow_emitters).

File: Content.Shared/CCVar/CCVars.cs [hunk 1]
```diff
@@ -2023,6 +2023,20 @@ public sealed class CCVars : CVars
...
+            CVarDef.Create("mime.allow_footsteps", true, CVar.REPLICATED | CVar.SERVER);
+
+        /// <summary>
+        /// Whether mimes can make sounds indirectly by interactions
+        /// with entities that use the EmitSound components.
+        /// This includes bike horns, desk bells, plushies, etc.
+        /// </summary>
+        public static readonly CVarDef<bool> MimeAllowEmitters =
+            CVarDef.Create("mime.allow_emitters", true, CVar.REPLICATED | CVar.SERVER);
+
```
Comment (metalgearsloth) [inline Content.Shared/CCVar/CCVars.cs:2039]: These should be component fields but unless a fork wants them then I would just drop them.

File: Content.Shared/Sound/EntitySystems/SilencedSystem.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
+using Content.Shared.Nutrition;
+using Content.Shared.Sound.Components;
+using Content.Shared.Tag;
+
+namespace Content.Shared.Sound.EntitySystems;
+
+public sealed partial class SilencedSystem : EntitySystem
+{
```
Comment (metalgearsloth) [inline Content.Shared/Sound/EntitySystems/SilencedSystem.cs:8]: Needs documenting.

File: Content.Shared/Sound/EntitySystems/SilencedSystem.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        if (!entity.Comp.AllowEmitterUse)
+            args.Cancel();
+    }
+
+    private void OnComponentStartup(Entity<SilencedComponent> entity, ref ComponentStartup args)
+    {
+        if (!entity.Comp.AllowFootsteps)
+            entity.Comp.HadFootsteps = _tag.RemoveTag(entity, "FootstepSound");
+    }
+
```
Comment (metalgearsloth) [inline Content.Shared/Sound/EntitySystems/SilencedSystem.cs:38]: MapInit

File: Content.Shared/Sound/SharedEmitSoundSystem.cs [hunk 4]
```diff
@@ -145,3 +160,22 @@ private void OnEmitSoundOnCollide(EntityUid uid, EmitSoundOnCollideComponent com
...
+
+/// <summary>
+/// Raised on an entity using a sound-emitting object when it tries to emit a sound.
+/// Cancelling will prevent the sound from being emitted.
+/// </summary>
+[ByRefEvent]
+public sealed class UseSoundEmitterAttemptEvent(EntityUid usedEnt) : CancellableEntityEventArgs
+{
+    public readonly EntityUid UsedEnt = usedEnt;
+}
```
Comment (metalgearsloth) [inline Content.Shared/Sound/SharedEmitSoundSystem.cs:181]: record structs please.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Shared/CCVar/CCVars.cs (hunk 1 above)
Comment (Tayrtahn) [reply Content.Shared/CCVar/CCVars.cs:2039]: Looks like I missed this while cleaning up. I'll delete it when I get the chance.

**Comment by UbaserB** [comment]
can you re-add footsteps but make it quieter? Kind of like tiptoeing, y'know mimes probably have perfected the art of walking/running silently.

**Comment by Tayrtahn** [comment]
With snoring now a trait, the only part of this left is the silent eating/drinking, and there are better ways of implementing that by itself (like a component that can control an entity's eating/drinking sounds, instead of just silencing based on mime powers), so I'm going to close this.

**Comment by lzk228** [comment]
you probably still can do removing snoring component for mmime

**Pull request closed without merging**